File: app/src/main/java/fr/dmconcept/bob/models/Project.java
package fr.dmconcept.bob.models;

import android.util.Log;

import java.util.ArrayList;

public class Project {

    private static final String TAG = "models.Project";

    // The project id
    private Long mId;

    // The project name
    private String mName;

    // The project steps
    private ArrayList<Step> mSteps;

    // The project board config
    private BoardConfig  mBoardConfig;

    /**
     * Create a new project with a start and end step
     *
     * @param name the project mName
     * @param boardConfig the board config
     */
    public Project(String name, BoardConfig boardConfig) {

        this(-1, name, boardConfig, new ArrayList<Step>());

        addStep(2000);
        addStep(0);
    }

    /**
     * Create a new project
     *
     * @param id the project mId
     * @param name the project mName
     * @param boardConfig the project board configuration
     * @param steps the project mSteps (must contain at leat a start and end step)
     */
    public Project(long id, String name, BoardConfig boardConfig, ArrayList<Step> steps) {

        assert steps.size() >= 2;

        // Check that the all the steps contains the same servo
        // count as defined in the project's board config
        for(Step s: steps)
            //TODO ain't working nah?
            assert s.getPositions().size() == boardConfig.getServoCount();

        this.mId = id   ;
        this.mName = name ;
        this.mBoardConfig = boardConfig;
        this.mSteps = steps;
    }

    /**
     * @return the project id
     */
    public long getId() {
        return mId;
    }

    /**
     * @return the project name
     */
    public String getName() {
        return mName;
    }

    /**
     * @return the steps
     */
    public ArrayList<Step> getSteps() {
        return mSteps;
    }

    /**
     * Return a step
     * @param i the step index
     * @return the step
     */
    public Step getStep(int i) {
        return mSteps.get(i);
    }

    /**
     * @return the project board config
     */
    public BoardConfig getBoardConfig() {
        return mBoardConfig;
    }

    /**
     * @return the project duration in ms
     */
    public int getDuration() {

        int duration = 0;

        for (Step step : mSteps)
            duration += step.getDuration();

        return duration;

    }

    /**
     * Create a new step at the end of the project
     */
    public void addStep(int duration) {

        Log.i(TAG, "addStep(" + duration + ")");

        ArrayList<Integer> positions = new ArrayList<Integer>();

        for(ServoConfig ignored : mBoardConfig.getServoConfigs())
            positions.add(50);

        // Set the length on the last step
        mSteps.get(mSteps.size() - 1).setDuration(duration);

        // Add the new step
        mSteps.add(new Step(0, positions));

    }

    public void removeStep(int i) {

        Log.i(TAG, "removeStep(" + i + ")");

        mSteps.remove(i);

        // If the last step is removed, set the last step duration to 0
        if (i == mSteps.size()) {
            Log.i(TAG, "removeStep() - Removing last step setting the new last step duration to 0");
            mSteps.get(mSteps.size() -1).setDuration(0);
        }

    }

}

File: bob-client/src/main/scala/fr/dmconcept/bob/client/models/Project.scala
package fr.dmconcept.bob.client.models

import java.util.UUID

object Project extends BobSerializable[Project] {

  final val TAG = "models.Project"

  /**
   * The step default start and end position when creating an empty project
   */
  final val STEP_DEFAULT_POSITION = (0, 100)

  def serialize(project: Project) = project.serialize

  def deserialize(serialized: Map[String, Any]): Project = {

    val id          = serialized("id"  ).asInstanceOf[String]
    val name        = serialized("name").asInstanceOf[String]
    val boardConfig = BoardConfig.deserialize(serialized("boardConfig").asInstanceOf[Map[String, Any]])
    val steps       = serialized("steps").asInstanceOf[List[Map[String, Any]]].map(Step.deserialize).toVector

    Project(id, name, boardConfig, steps)

  }

  /**
   * Creates a new Project with a start and end Step
   *
   * @param name the project name
   * @param boardConfig the board config of the project
   *
   * @return the project
   */
  def apply(name: String, boardConfig: BoardConfig): Project = {

    val randomUUID = UUID.randomUUID.toString

    // Create the start and end steps with the same position count as defined
    // in the board config
    val positionsCount = boardConfig.servoConfigs.length

    val steps: Vector[Step] = Vector(
      Step(   Vector.fill(positionsCount)(  0)),
      Step(0, Vector.fill(positionsCount)(100))
    )

    Project(randomUUID, name, boardConfig, steps)

  }

}

case class Project(

  id: String,

  name: String,

  boardConfig: BoardConfig,

  steps: Vector[Step]

) {

  assert(!name.isEmpty           , "The project name is empty"                )
  assert(steps.length >=2        , "The project must contain at least 2 steps")
  assert(steps.last.duration == 0, "The last step duration must be equal to 0")

  steps.foreach(checkPositionsCount)

  /**
   * @return the project duration in seconds
   */
  def duration: Int = steps.map(_.duration).sum

  /*
  /**
   * Create a new step at the end of the project
   */
  public void addStep(int duration) {

      Log.i(TAG, "addStep(" + duration + ")");

      ArrayList<Integer> positions = new ArrayList<Integer>();

      for(ServoConfig ignored : mBoardConfig.getServoConfigs())
          positions.add(50);

      // Set the length on the last step
      mSteps.get(mSteps.size() - 1).setDuration(duration);

      // Add the new step
      mSteps.add(new Step(0, positions));

  }

  public void removeStep(int i) {

      Log.i(TAG, "removeStep(" + i + ")");

      mSteps.remove(i);

      // If the last step is removed, set the last step duration to 0
      if (i == mSteps.size()) {
          Log.i(TAG, "removeStep() - Removing last step setting the new last step duration to 0");
          mSteps.get(mSteps.size() -1).setDuration(0);
      }

  }
  */

  /**
   * Check that the step contains the same number of positions as defined in the board config
   */
  private def checkPositionsCount(step: Step) {

    val expected = boardConfig.servoConfigs.length

    assert(
      step.positions.length == expected,
      s"The step position count don't match the board config servo count: expecting ${expected} but got ${step.positions.length} in $step"
    )

  }

  def serialize = Map(
    "id"          -> id,
    "name"        -> name,
    "boardConfig" -> boardConfig.serialize,
    "steps"       -> steps.map(Step.serialize)
  )

}
